d2ir/d2ir.go

```diff
@@ -0,0 +1,1046 @@
+		References: []*EdgeReference{{
+			Context: refctx,
+		}},
+	}
+	m.Edges = append(m.Edges, e)
+
+	return e, nil
+}
+
+func (s *Scalar) ast() d2ast.Node {
+	return s.Value
+}
+
+func (f *Field) ast() d2ast.Node {
+	k := &d2ast.Key{
+		Key: &d2ast.KeyPath{
+			Path: []*d2ast.StringBox{
+				d2ast.MakeValueBox(d2ast.RawString(f.Name, true)).StringBox(),
+			},
+		},
+	}
+
+	if f.Primary_ != nil {
+		k.Primary = d2ast.MakeValueBox(f.Primary_.ast().(d2ast.Value)).ScalarBox()
+	}
+	if f.Composite != nil {
+		k.Value = d2ast.MakeValueBox(f.Composite.ast().(d2ast.Value))
+	}
+
+	return k
+}
+
+func (e *Edge) ast() d2ast.Node {
+	astEdge := &d2ast.Edge{}
+
+	astEdge.Src = d2ast.MakeKeyPath(e.ID.SrcPath)
+	if e.ID.SrcArrow {
+		astEdge.SrcArrow = "<"
+	}
+	astEdge.Dst = d2ast.MakeKeyPath(e.ID.DstPath)
+	if e.ID.DstArrow {
+		astEdge.DstArrow = ">"
+	}
+
+	k := &d2ast.Key{
+		Edges: []*d2ast.Edge{astEdge},
+	}
+
+	if e.Primary_ != nil {
+		k.Primary = d2ast.MakeValueBox(e.Primary_.ast().(d2ast.Value)).ScalarBox()
+	}
+	if e.Map_ != nil {
+		k.Value = d2ast.MakeValueBox(e.Map_.ast().(*d2ast.Map))
+	}
+
+	return k
+}
+
+func (a *Array) ast() d2ast.Node {
+	if a == nil {
+		return nil
+	}
+	astArray := &d2ast.Array{}
+	for _, av := range a.Values {
+		astArray.Nodes = append(astArray.Nodes, d2ast.MakeArrayNodeBox(av.ast().(d2ast.ArrayNode)))
+	}
+	return astArray
+}
+
+func (m *Map) ast() d2ast.Node {
+	if m == nil {
+		return nil
+	}
+	astMap := &d2ast.Map{}
+	if m.Root() {
+		astMap.Range = d2ast.MakeRange(",0:0:0-1:0:0")
+	} else {
+		astMap.Range = d2ast.MakeRange(",1:0:0-2:0:0")
+	}
+	for _, f := range m.Fields {
+		astMap.Nodes = append(astMap.Nodes, d2ast.MakeMapNodeBox(f.ast().(d2ast.MapNode)))
+	}
+	for _, e := range m.Edges {
+		astMap.Nodes = append(astMap.Nodes, d2ast.MakeMapNodeBox(e.ast().(d2ast.MapNode)))
+	}
+	return astMap
+}
+
+func (m *Map) appendFieldReferences(i int, kp *d2ast.KeyPath, refctx *RefContext) {
+	sb := kp.Path[i]
+	f := m.GetField(sb.Unbox().ScalarString())
+	if f == nil {
+		return
+	}
+
+	f.References = append(f.References, &FieldReference{
+		String:  sb.Unbox(),
+		KeyPath: kp,
+		Context: refctx,
+	})
+	if i+1 == len(kp.Path) {
+		return
+	}
+	if f.Map() != nil {
+		f.Map().appendFieldReferences(i+1, kp, refctx)
+	}
+}
+
+func ParentMap(n Node) *Map {
+	for {
+		n = n.Parent()
+		if n == nil {
+			return nil
+		}
+		if m, ok := n.(*Map); ok {
+			return m
+		}
+	}
+}
+
+func ParentField(n Node) *Field {
+	for {
+		n = n.Parent()
+		if n == nil {
+			return nil
+		}
+		if f, ok := n.(*Field); ok {
+			return f
+		}
+	}
+}
+
+func ParentLayer(n Node) Node {
+	for {
+		n = n.Parent()
+		if n == nil {
+			return nil
+		}
+		if NodeLayerKind(n) != "" {
+			return n
+		}
+	}
+}
+
+func ParentEdge(n Node) *Edge {
+	for {
+		n = n.Parent()
+		if n == nil {
+			return nil
+		}
+		if e, ok := n.(*Edge); ok {
+			return e
+		}
+	}
+}
+
+func countUnderscores(p []string) int {
+	var count int
+	for _, el := range p {
```

Comment:
for i, el

and then return i+1 instead of break

Reply:
I'd prefer to keep as is for clarity. p can never be passed in as nil but if it were, I don't believe your suggestion would work. As you couldn't check whether the loop passed all iterations or if not a single without checking the length of p.